### ISSUE TITLE
After upgrading opam libraries: make inconsistent assumptions over interface

### ISSUE THREAD

**Issue by copy** [ocaml/dune#1893]
I just upgraded tls and got a spurious "make inconsistent assumptions over interface Tls" error. It goes away after `dune clean`.

To reproduce, create these files:

```
(library (name lib) (modules lib) (libraries tls))
(executable (name app) (modules app) (libraries lib))
```

<hr>

```ocaml
(* app.ml *)
Lib.main ()
```

<hr>

```ocaml
(* lib.ml *)
let main () = ()
let _ = Tls.Engine.server
```

<hr>

```ocaml
(* lib.mli *)
val main : unit -> unit
```

And run the following steps:

- `opam install --yes tls=0.9.3`
- `dune build app.exe`
- `opam install --yes tls=0.10.1`
- `dune build app.exe`

Tested on dune 1.7.3.

**Comment by diml** [comment]
I just tested with master and I can't reproduce this error. Could you double check your reproduction case?

**Comment by copy** [comment]
I can reproduce this on a clean 4.07.1 switch with dune pinned to the repo. I have attached a log of running `dune build --verbose app.exe`.

[dune.log](https://github.com/ocaml/dune/files/2927317/dune.log)

**Comment by diml** [comment]
I'm wondering if it's not due to the dune-package changes. tls doesn't seem to use dune. It is possible that we miss these dependencies for non-dune packages (/cc @rgrinberg)

**Comment by rgrinberg** [comment]
Okay, I will try to reproduce this with dune 1.6. It's possible that dune is somehow dropping an archive somewhere in the META -> dune-package conversion for example.

**Comment by copy** [comment]
Fixed in #1965.
